File: files/CH-038.R
library(tidyverse)
library(readxl)

input = read_excel("files/CH-038 Duration Since Last Visit.xlsx", range = "B2:C26")
test  = read_excel("files/CH-038 Duration Since Last Visit.xlsx", range = "G2:H6")

dates = seq(as.Date("2024-01-01"), as.Date("2024-04-01"), by = "month") %>%
  as_tibble() %>%
  mutate(end_of_month = value + months(1) - days(1)) %>%
  select(end_of_month) 

ends = expand_grid(Date = dates$end_of_month, `Agent ID` = unique(input$`Agent ID`)) %>%
  mutate(type = "end")

result = input %>%
  mutate(type = "visit") %>%
  bind_rows(ends) %>%
  arrange(`Agent ID`, Date) %>%
  group_by(`Agent ID`) %>%
  mutate(last_visit = if_else(type == "visit", as.Date(as.POSIXct(Date)), NA)) %>%
  fill(last_visit, .direction = "down") %>%
  mutate(month = month(Date)) %>%
  filter(type == "end") %>%
  mutate(datediff = difftime(Date, last_visit, units = "days") %>% as.numeric()) %>%
  ungroup() %>%
  summarise(mean = mean(datediff, na.rm = TRUE), .by = "month")

identical(result$mean, test$`AVG Duration from Last Visit`)
# [1] TRUE

File: files/000-099/038/CH-038.py
import pandas as pd
from datetime import datetime, timedelta

input = pd.read_excel("CH-038 Duration Since Last Visit.xlsx", usecols="B:C", skiprows=1, nrows= 25)
test = pd.read_excel("CH-038 Duration Since Last Visit.xlsx", usecols="G:H", skiprows=1, nrows = 4)

dates = pd.date_range(start="2024-01-01", end="2024-05-01", freq="M").to_frame(name="end_of_month")

ends = pd.MultiIndex.from_product([dates["end_of_month"], input["Agent ID"].unique()], names=["Date", "Agent ID"]).to_frame(index=False)
ends["type"] = "end"

result = pd.concat([input.assign(type="visit"), ends]).sort_values(by=["Agent ID", "Date"])
result["last_visit"] = result["Date"].where(result["type"] == "visit").groupby(result["Agent ID"]).ffill()
result["month"] = result["Date"].dt.month.astype("int64")
result = result[result["type"] == "end"]
result["datediff"] = (result["Date"] - result["last_visit"]).dt.days
result = result.groupby("month")["datediff"].mean().reset_index()

result.columns = ["Month", "AVG Duration from Last Visit"]
print(result.equals(test)) # True